File: app/src/main/java/apps/amine/bou/readerforselfoss/api/selfoss/SelfossService.java
package apps.amine.bou.readerforselfoss.api.selfoss;


import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;


interface SelfossService {
    @GET("login")
    Call<SuccessResponse> loginToSelfoss(@Query("username") String username, @Query("password") String password);

    @GET("items")
    Call<List<Item>> getItems(@Query("type") String type, @Query("username") String username, @Query("password") String password);

    @POST("mark/{id}")
    Call<SuccessResponse> markAsRead(@Path("id") String id, @Query("username") String username, @Query("password") String password);


    @POST("unmark/{id}")
    Call<SuccessResponse> unmarkAsRead(@Path("id") String id, @Query("username") String username, @Query("password") String password);

    @FormUrlEncoded
    @POST("mark")
    Call<SuccessResponse> markAllAsRead(@Field("ids[]") List<String> ids, @Query("username") String username, @Query("password") String password);


    @POST("starr/{id}")
    Call<SuccessResponse> starr(@Path("id") String id, @Query("username") String username, @Query("password") String password);


    @POST("unstarr/{id}")
    Call<SuccessResponse> unstarr(@Path("id") String id, @Query("username") String username, @Query("password") String password);


    @GET("stats")
    Call<Stats> stats(@Query("username") String username, @Query("password") String password);


    @GET("tags")
    Call<List<Tag>> tags(@Query("username") String username, @Query("password") String password);


    @GET("update")
    Call<String> update(@Query("username") String username, @Query("password") String password);

    @GET("sources/spouts")
    Call<Map<String, Spout>> spouts(@Query("username") String username, @Query("password") String password);

    @GET("sources/list")
    Call<List<Sources>> sources(@Query("username") String username, @Query("password") String password);


    @DELETE("source/{id}")
    Call<SuccessResponse> deleteSource(@Path("id") String id, @Query("username") String username, @Query("password") String password);

    @FormUrlEncoded
    @POST("source")
    Call<SuccessResponse> createSource(@Field("title") String title, @Field("url") String url, @Field("spout") String spout, @Field("tags") String tags, @Field("filter") String filter, @Query("username") String username, @Query("password") String password);
}

File: app/src/main/java/apps/amine/bou/readerforselfoss/api/selfoss/SelfossService.kt
package apps.amine.bou.readerforselfoss.api.selfoss


import retrofit2.Call
import retrofit2.http.DELETE
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query


internal interface SelfossService {
    @GET("login")
    fun loginToSelfoss(@Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>

    @GET("items")
    fun getItems(@Query("type") type: String,
                 @Query("tag") tag: String?,
                 @Query("source") source: Long?,
                 @Query("username") username: String,
                 @Query("password") password: String): Call<List<Item>>

    @POST("mark/{id}")
    fun markAsRead(@Path("id") id: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>


    @POST("unmark/{id}")
    fun unmarkAsRead(@Path("id") id: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>

    @FormUrlEncoded
    @POST("mark")
    fun markAllAsRead(@Field("ids[]") ids: List<String>, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>


    @POST("starr/{id}")
    fun starr(@Path("id") id: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>


    @POST("unstarr/{id}")
    fun unstarr(@Path("id") id: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>


    @GET("stats")
    fun stats(@Query("username") username: String, @Query("password") password: String): Call<Stats>


    @GET("tags")
    fun tags(@Query("username") username: String, @Query("password") password: String): Call<List<Tag>>


    @GET("update")
    fun update(@Query("username") username: String, @Query("password") password: String): Call<String>

    @GET("sources/spouts")
    fun spouts(@Query("username") username: String, @Query("password") password: String): Call<Map<String, Spout>>

    @GET("sources/list")
    fun sources(@Query("username") username: String, @Query("password") password: String): Call<List<Sources>>


    @DELETE("source/{id}")
    fun deleteSource(@Path("id") id: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>

    @FormUrlEncoded
    @POST("source")
    fun createSource(@Field("title") title: String, @Field("url") url: String, @Field("spout") spout: String, @Field("tags") tags: String, @Field("filter") filter: String, @Query("username") username: String, @Query("password") password: String): Call<SuccessResponse>
}
